clang/lib/Sema/SemaTemplate.cpp

```diff
@@ -7991,8 +7991,12 @@ Sema::CheckTemplateDeclScope(Scope *S, TemplateParameterList *TemplateParams) {
   //   have C linkage.
   DeclContext *Ctx = S->getEntity();
   if (Ctx && Ctx->isExternCContext()) {
-    Diag(TemplateParams->getTemplateLoc(), diag::err_template_linkage)
-        << TemplateParams->getSourceRange();
+    SourceRange Range =
+        TemplateParams->getTemplateLoc().isInvalid() && TemplateParams->size()
+            ? TemplateParams->getParam(TemplateParams->size() - 1)
+                  ->getSourceRange()
```

Comment:
Why use the last one and not the first one?
(Either getTemplateLoc is valid or all parameters are invented)

Reply:
@cor3ntin thanks for the review. The decision to retrieve the last parameter was driven by the following comment for this helper https://github.com/llvm/llvm-project/blob/36cd60144b15dd35c8e0081100421c3511242e02/clang/include/clang/AST/DeclTemplate.h#L1091

Or better to use first param?

Reply:
I think using the first param would be better, yes

(in the general case, you can have

```cpp
template <typename>
void f(auto);
```

so to establish if there are abbreviated parameters, you need to look at the first one, 
(and getTemplateLoc is always valid)

but in the 
```cpp
void f(auto)
```

case, there are only abbreviated parameters, and you can point to the first one (and getTemplateLoc is never valid).

Does that make sense?